### PR TITLE
Review chap 6

### PR DESCRIPTION
**Pull request by maiq721**

### CODE REVIEW

File: app/models/user.rb [hunk 1]
```diff
@@ -0,0 +1,8 @@
+class User < ApplicationRecord
+  before_save { self.email = email.downcase }
```
Comment (mdmahmudurrahman) [inline app/models/user.rb]: Convention: Do not add white spaces before and after [] () {}

File: app/models/user.rb [hunk 2]
```diff
@@ -0,0 +1,8 @@
...
+  validates :name, presence: true, length: { maximum: 50 }
+  VALID_EMAIL_REGEX = /\A[\w+\-.]+@[a-z\d\-]+(\.[a-z\d\-]+)*\.[a-z]+\z/i
+  validates :email, presence: true, length: { maximum: 255 }, format: { with: VALID_EMAIL_REGEX }, uniqueness: { case_sensitive: false }
```
Comment (mdmahmudurrahman) [inline app/models/user.rb]: 1. Do not add white spaces before and after [] () {}
2. Maximum 80 characters per line

File: app/models/user.rb [hunk 3]
```diff
@@ -0,0 +1,10 @@
+class User < ApplicationRecord
+  before_save{ self.email = email.downcase }
```
Comment (mdmahmudurrahman) [inline app/models/user.rb]: No. Convention: Do not add white spaces before and after [] () {} means the code like following 
```
before_save {self.email = email.downcase}
```

File: app/models/user.rb [hunk 4]
```diff
@@ -0,0 +1,10 @@
...
+  validates :name, presence: true, length:{ maximum: 50 }
```
Comment (mdmahmudurrahman) [inline app/models/user.rb]: same here

File: app/models/user.rb [hunk 5]
```diff
@@ -0,0 +1,10 @@
...
+  VALID_EMAIL_REGEX = /\A[\w+\-.]+@[a-z\d\-]+(\.[a-z\d\-]+)*\.[a-z]+\z/i
+  validates :email, presence: true, length:{ maximum: 255 },
+   format:{ with: VALID_EMAIL_REGEX }, 
+   uniqueness:{ case_sensitive: false }
+  has_secure_password
+  validates :password, presence: true, length:{ minimum: 6 }
```
Comment (mdmahmudurrahman) [inline app/models/user.rb]: In this file, you can see many constants. Please use "Config" gem to manage the constants 
Reference: [https://github.com/railsconfig/config](https://github.com/railsconfig/config)

File: test/fixtures/users.yml [hunk 6]
```diff
@@ -0,0 +1 @@
+#empty
```
Comment (mdmahmudurrahman) [inline test/fixtures/users.yml]: Please delete this test file

File: test/models/user_test.rb [hunk 7]
```diff
@@ -0,0 +1,7 @@
+require "test_helper"
```
Comment (mdmahmudurrahman) [inline test/models/user_test.rb]: Please delete this test file

File: app/models/user.rb [hunk 8]
```diff
@@ -0,0 +1,10 @@
+class User < ApplicationRecord
+  before_save{ self.email = email.downcase }
+  validates :name, presence: true, length:{ maximum: 50 }
+  VALID_EMAIL_REGEX = /\A[\w+\-.]+@[a-z\d\-]+(\.[a-z\d\-]+)*\.[a-z]+\z/i
+  validates :email, presence: true, length:{ maximum: 255 },
+   format:{ with: VALID_EMAIL_REGEX },
```
Comment (mdmahmudurrahman) [inline app/models/user.rb]: 2 space indentation

File: app/models/user.rb [hunk 9]
```diff
@@ -0,0 +1,10 @@
+class User < ApplicationRecord
+  before_save{ self.email = email.downcase }
+  validates :name, presence: true, length:{ maximum: 50 }
+  VALID_EMAIL_REGEX = /\A[\w+\-.]+@[a-z\d\-]+(\.[a-z\d\-]+)*\.[a-z]+\z/i
+  validates :email, presence: true, length:{ maximum: 255 },
+   format:{ with: VALID_EMAIL_REGEX }, 
+   uniqueness:{ case_sensitive: false }
```
Comment (mdmahmudurrahman) [inline app/models/user.rb]: same here

File: app/assets/stylesheets/custom.scss [hunk 10]
```diff
@@ -2,6 +2,8 @@
 @import 'bootstrap';
 @import 'define_color.scss';
 
+$gray-medium-light: #eaeaea;
```
Comment (mdmahmudurrahman) [inline app/assets/stylesheets/custom.scss]: Please move this variable to the external file where you declare the color variable

File: app/models/user.rb [hunk 11]
```diff
@@ -0,0 +1,10 @@
+class User < ApplicationRecord
+  before_save {self.email = email.downcase}
```
Comment (mdmahmudurrahman) [inline app/models/user.rb]: Can I ask you a question: What is before_save in rails

File: app/models/user.rb [hunk 12]
```diff
@@ -0,0 +1,10 @@
...
+  validates :name, presence: true, length: {maximun: Settings.validate.name.maximum}
+  VALID_EMAIL_REGEX = /\A[\w+\-.]+@[a-z\d\-]+(\.[a-z\d\-]+)*\.[a-z]+\z/i
+  validates :email, presence: true, length: {maximum: Settings.validate.email.maximum},
+    format: {with: VALID_EMAIL_REGEX},
+    uniqueness: {case_sensitive: false}
+  has_secure_password
```
Comment (mdmahmudurrahman) [inline app/models/user.rb]: I have another question: What is has_secure_password and why you used it here?

File: app/models/user.rb (hunk 11 above)
Comment (maiq721) [reply app/models/user.rb]: before_save callback the method before save data in table. in this case , It downcases email before save user email to standardize all email

File: app/models/user.rb (hunk 12 above)
Comment (maiq721) [reply app/models/user.rb]: It can validate password automatically. I used it here to validate password for User

Comment (mdmahmudurrahman) [reply app/models/user.rb]: @maiq721 I am not satisfied about the answer. Could you please study more and add more words to your answer.

Comment (maiq721) [reply app/models/user.rb]: I got it correctlly... has_secure_password is the method encrypt password with Bcrypt, add a password attribute(read-only) to the model,ensure that password_digest is present before model is saved, validates that the password is confirmed(you have to confirm password when you register an account on some web site) and authenticate password. I thought all of them are validate.

**Pull request merged**